src/components/shape-helper.js

```diff
@@ -0,0 +1,81 @@
+import * as threeToAmmo from "three-to-ammo";
+import { SHAPE, FIT } from "three-ammo/constants";
+
+AFRAME.registerComponent("shape-helper", {
+  schema: {
+    type: {
+      default: SHAPE.HULL,
+      oneOf: [
+        SHAPE.BOX,
+        SHAPE.CYLINDER,
+        SHAPE.SPHERE,
+        SHAPE.CAPSULE,
+        SHAPE.CONE,
+        SHAPE.HULL,
+        SHAPE.HACD,
+        SHAPE.VHACD,
+        SHAPE.MESH,
+        SHAPE.HEIGHTFIELD
+      ]
+    },
+    fit: { default: FIT.ALL, oneOf: [FIT.ALL, FIT.MANUAL] },
+    halfExtents: { type: "vec3", default: { x: 1, y: 1, z: 1 } },
+    minHalfExtent: { default: 0 },
+    maxHalfExtent: { default: Number.POSITIVE_INFINITY },
+    sphereRadius: { default: NaN },
+    cylinderAxis: { default: "y", oneOf: ["x", "y", "z"] },
+    margin: { default: 0.01 },
+    offset: { type: "vec3", default: { x: 0, y: 0, z: 0 } },
+    orientation: { type: "vec4", default: { x: 0, y: 0, z: 0, w: 1 } },
+    heightfieldData: { default: [] },
+    heightfieldDistance: { default: 1 },
+    includeInvisible: { default: false }
+  },
+
+  multiple: true,
+
+  init: function() {
+    this.system = this.el.sceneEl.systems["hubs-systems"].physicsSystem;
+    this.system.registerShapeHelper(this);
+  },
+
+  init2: function() {
+    this.mesh = null;
+
+    let bodyEl = this.el;
+    this.bodyHelper = bodyEl.components["body-helper"] || null;
+    while (!this.bodyHelper && bodyEl.parentNode != this.el.sceneEl) {
+      bodyEl = bodyEl.parentNode;
+      if (bodyEl.components["body-helper"]) {
+        this.bodyHelper = bodyEl.components["body-helper"];
+      }
+    }
+    if (!this.bodyHelper) {
+      console.warn("body not found");
+      return;
+    }
+    if (this.data.fit !== FIT.MANUAL) {
+      if (!this.el.object3DMap.mesh) {
+        console.error("Cannot use FIT.ALL without object3DMap.mesh");
+        return;
+      }
+      this.mesh = this.el.object3DMap.mesh;
+      this.mesh.updateMatrices();
+    }
+
+    this.shapes = threeToAmmo.createCollisionShapes(this.mesh, this.data);
+    for (let i = 0; i < this.shapes.length; i++) {
+      this.bodyHelper.body.addShape(this.shapes[i]);
+    }
+  },
+
+  remove: function() {
+    if (this.shapes) {
+      for (let i = 0; i < this.shapes.length; i++) {
+        if (this.bodyHelper.body) {
+          this.bodyHelper.body.removeShape(this.shapes[i]);
```

Comment:
I think it's very confusing that three-to-ammo's `removeShape` also destroys the shape. It's very not-RAII, it looks like a bug, and it prevents you from reusing a shape in multiple bodies, which it seems like you could otherwise do.

Reply:
> 
> 
> I don't think the body/shape helper "registration" code is correct. Consider the following case:
> 
>     1. Ammo starts loading.
> 
>     2. A body helper component is initialized. `registerBodyHelper` is called.
> 
>     3. The body helper component is removed. `removeBody` is called.
> 
>     4. Ammo finishes loading. `world` is created.
> 
>     5. A physics system tick occurs. `init2` is called. `addBody` is called. This body will be leaked forever.
> 
> 
> If you swap the order of 3 and 4, the same result happens.
> 
> Also:
> 
>     1. Ammo starts loading.
> 
>     2. Body helper component X is initialized. `registerBodyHelper` is called.
> 
>     3. Ammo finishes loading. `world` is created.
> 
>     4. Body helper component Y is initialized. `registerBodyHelper` is called. `world` is non-null, so `init2` on Y is called directly.
> 
>     5. A physics system tick occurs. `init2` on X is called.
> 
> 
> This violated the invariant that bodies are initialized in the order the components are created. I don't know if that creates any problems, but it's certainly a little surprising.
> 
> There are many ways you can fix this, but here is how I believe I would have written the code:
> 
>     * Store the promise returned by `AmmoModule().then(...)` in the physics system.
> 
>     * Create a physics system function `whenReady` which returns that promise.
> 
>     * Remove the `bodyHelpers` and `shapeHelpers` arrays in the physics system, and the `registerBodyHelper` and `registerShapeHelper` functions.
> 
>     * In the body and shape helper components, write something like:
> 
> 
> ```js
> init: function() {
>   physicsSystem.whenReady().then(world => {
>     if (!this.removed) {
>       this.body = world.addBody(new Body(/* ... */));
>     }
>   });
> }
> remove: function() {
>   this.removed = true;
>   if (this.body != null) {
>     physicsSystem.removeBody(this.body);
>   }
> }
> ```
> 
> (The shape helper one is more complicated, but I don't see any reason it should be a problem.)

You bring up a good point about what happens if these helper components get destroyed before world is created. In practice, this never really happens because the ammo .wasm loads very quickly before any of these components get initialized- this is all just trying to avoid any issues that might come in the case where something else terrible has gone wrong and the wasm somehow doesn't load at all.

That said, there is a reason I didn't do this with promises as you suggested:
1. The promise returned by `AmmoModule()` is not a "real" promise. This is not a huge deal because you can create a "real" promise and return that instead, and have at promise resolved when this fake one does.
2. It is important that bodies are initialized before shapes do, since they have to add themselves to the body.

I think though that in practice what will happen is you will get an error because you will try to call `tick2` on a component that no longer exists. So I think the correct fix is to check to make sure the "helper" component is still valid. 

As for the second concern about the order in which `init2` is called on bodies, it doesn't really matter in practice. However, the problem is resolved if the iteration of `this.bodyHelpers` and `this.shapeHelpers` is done in the `AmmoModule().then()` promise as opposed to in `tick` as you  suggested in another comment.

Reply:
I see, the issue of ordering between the bodies and their corresponding shapes does make it a lot less straightforward, and it's not obvious to me anymore how to improve on the way you did it (besides fixing correctness with some more explicit checks as you suggested.) Maybe I will think about it a little more.

Reply:
(My first thought is that if you care a lot about the initialization ordering of separate components, you probably shouldn't have modelled them as A-Frame components. Components are supposed to be kind of conceptually independent of each other with independent lifecycles. So maybe you could remove the shape component and implement shapes by calling a function on the body component instead.)